### PR TITLE
174785971 required navigation 2

### PR DESCRIPTION
**Pull request by mklewandowski**
This PR adds handling of required questions on an activity page.  If one or more incomplete required questions are found on a page, forward page navigation is prevented.  If in this case forward navigation is selected, a message is shown to the user requesting that they complete the questions first.  

Changes include:
- add `navigation` listener in `iFrameRuntime` component to obtain required status of questions on page
- store `incompleteQuestions` in local state of `App` component to determine if we have required questions which have not been completed
- disable navigation components if required questions are incomplete
- rework the page buttons in the `NavPages` component so we have more control over button state
- add `ModalDialog` component that can show a modal dialog with a label and close button.  This control can be reused and expanded for future cases where we need a modal dialog
- update and expand sample data

PR can be tested on this branch or locally using the default sample activity (page 5 contains multiple required questions):
https://activity-player.concord.org/branch/required-navigation-2/?preview&page=5

There are still some tests that I want to add to this PR, but I wanted folks to be able to take a look at the PR while I work on tests.

### CODE REVIEW

File: src/components/activity-page/managed-interactive/iframe-runtime.tsx [hunk 1]
```diff
@@ -179,7 +185,7 @@ export const IframeRuntime: React.FC<IProps> = forwardRef((props, ref) => {
       }
     };
   }, [url, authoredState, report, initialInteractiveState, setNewHint, getFirebaseJWT, setSupportedFeatures,
-      setSendCustomMessage, showModal, closeModal]);
+      setSendCustomMessage, showModal, closeModal, setNavigation]);
```
Comment (kswenson) [inline src/components/activity-page/managed-interactive/iframe-runtime.tsx:188]: There will be a trivial conflict here with my outstanding PR #74.

File: src/components/app.tsx [hunk 2]
```diff
@@ -230,8 +256,13 @@ export class App extends React.PureComponent<IProps, IState> {
...
   private handleChangePage = (page: number) => {
-    this.setState({currentPage: page});
-    setDocumentTitle(this.state.activity, page);
+    if (page > this.state.currentPage && this.state.incompleteQuestions.length > 0) {
+      const label = this.state.incompleteQuestions[0].navOptions?.message || kDefaultIncompleteMessage;
+      this.setShowModal(true, label);
+    } else if (page >= 0 && (this.state.activity && page <= this.state.activity.pages.length)) {
+      this.setState({currentPage: page, incompleteQuestions: []});
+      setDocumentTitle(this.state.activity, page);
+    }
```
Comment (kswenson) [inline src/components/app.tsx:265]: You could get rid of a bunch of `this.state`s above with:
```typescript
  const { currentPage, incompleteQuestions, activity } = this.state;
```

**Review by kswenson — COMMENTED** [review]
I know you haven't asked for review yet, but I took a quick read through and had a few suggestions along the way.

File: src/components/app.tsx [hunk 3]
```diff
@@ -22,10 +22,18 @@ import { TeacherEditionBanner }  from "./teacher-edition-banner";
...
 
 import "./app.scss";
 
 const kDefaultActivity = "sample-activity-multiple-layout-types";   // may eventually want to get rid of this
+const kDefaultIncompleteMessage = "Please submit an answer first.";
+
+interface IncompleteQuestion {
+  refId: string;
+  navOptions: INavigationOptions;
+}
```
[the remark this answers is not in the document]
Comment (mklewandowski) [reply src/components/app.tsx:37]: I think I fixed the redundancy a little earlier. Might want to refresh or pull latest.

File: src/components/activity-header/nav-pages.tsx [hunk 4]
```diff
@@ -1,46 +1,91 @@
...
+    const maxPagesRightOfCurrent = totalPages - currentPage;
+    let minPage = 1;
+    let maxPage = totalPages;
+    if (maxPagesLeftOfCurrent < 5) {
+      maxPage = Math.min(totalPages, currentPage + 5 + (5 - maxPagesLeftOfCurrent));
+    } else if (maxPagesRightOfCurrent < 5) {
+      minPage = Math.max(1, currentPage - 5 - (5 - maxPagesRightOfCurrent));
+    } else if (totalPages > kMaxPages) {
+      minPage = currentPage - 5;
+      maxPage = currentPage + 5;
```
Comment (kswenson) [inline src/components/activity-header/nav-pages.tsx:65]: Seems like all of these `5`s could be a constant like `kMaxButtonsPerSide`, which is probably computable as Math.floor(`kMaxNavigationButtons / 2)`. Then if we ever wanted to change the max number of navigation buttons everything should just work.

File: src/components/activity-page/managed-interactive/iframe-runtime.tsx (hunk 1 above)
Comment (mklewandowski) [reply src/components/activity-page/managed-interactive/iframe-runtime.tsx:188]: Yeah, I was assuming so after reviewing your earlier PR.

**Review by kswenson — APPROVED** [review]
👍 One more comment -- all pretty minor.

File: src/components/modal-dialog.tsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import React from "react";
+import Modal from "react-modal";
+
+import "./modal-dialog.scss";
+
+interface IProps {
+  title?: string;
+  label: string;
+  setShowModal: (show: boolean) => void;
```
[the remark this answers is not in the document]
Comment (mklewandowski) [reply src/components/modal-dialog.tsx]: Fixed.

File: src/components/app.tsx (hunk 2 above)
Comment (mklewandowski) [reply src/components/app.tsx:266]: Fixed.

File: src/components/activity-header/nav-pages.tsx [hunk 6]
```diff
@@ -1,46 +1,90 @@
 import React from "react";
 import IconHome from "../../assets/svg-icons/icon-home.svg";
-import ReactPaginate from "react-paginate";
+import { Page } from "../../types";
 
 import "./nav-pages.scss";
-import { Page } from "../../types";
+
+const kMaxPages = 11;
```
[the remark this answers is not in the document]
Comment (mklewandowski) [reply src/components/activity-header/nav-pages.tsx]: Fixed.

File: src/components/activity-header/nav-pages.tsx (hunk 4 above)
Comment (mklewandowski) [reply src/components/activity-header/nav-pages.tsx]: Good call.  Fixed.

**Pull request merged**